### PR TITLE
InkCanvas: report automation bounds and offscreen state correctly

### PR DESCRIPTION
**Pull request by Nitin-100**
## Problem

`InkCanvas` reports an **empty automation bounding rectangle** and `IsOffscreen = true` even when it is fully visible on screen. Assistive technology therefore skips the canvas: Narrator cannot navigate to it, Magnifier cannot track it, and UIA hit-testing and automated UI tests cannot locate it.

Measured in a WinUI 3 app with three canvases - one small and unclipped, two larger ones inside `ScrollViewer`s:

```
InkCanvas A (in scroller)    rect=EMPTY/INFINITE   offscreen=True
InkCanvas B (in scroller)    rect=EMPTY/INFINITE   offscreen=True
InkCanvas   (unclipped)      rect=EMPTY/INFINITE   offscreen=True
ScrollViewer / InkToolbar / TextBlock / Button    all correct, offscreen=False
```

Every other element in the same window reports correctly, including the canvases' own parents, so this is specific to `InkCanvas`.

It is not a missing peer and not a layout failure. Measured from inside the process:

```
InkCanvas: Actual=160x44 Visible Peer=InkCanvasAutomationPeer Rect=0,0 0x0      Offscreen=True
InkToolbar: Actual=204x44 Visible Peer=InkToolbarAutomationPeer Rect=21,140 294x54 Offscreen=False
```

The element has a real layout size and the correct peer attached, yet the rect is `0x0`.

## Cause

`InkCanvas` is a bare `FrameworkElement` with **no XAML children**. All of its content is a composition visual attached through `ElementCompositionPreview::SetElementChildVisual`.

The framework derives automation bounds from XAML render content. In the core, `CUIElement::GenerateContentBounds` returns an empty rect by default, and only core types override it (`CBorder`, `CPanel` when it has a brush, `CSwapChainPanel`). A hand-in composition visual is tracked separately and never reaches that path, so the canvas has no content bounds to measure - and `CFrameworkElementAutomationPeer::IsOffscreenHelper` treats empty bounds as offscreen.

## Fix

`InkCanvasAutomationPeer` derives the bounds from the element's *layout* size, then re-applies the clipping the framework would have applied, so both methods agree by construction:

- start from `ActualWidth`/`ActualHeight` transformed to root coordinates
- intersect with `UIElement.Clip` on the canvas **and** every ancestor, honouring the clip geometry's own `Transform` (the order the core uses in `CRectangle::InitializeRectangleClip`)
- intersect with the viewport of any `ScrollContentPresenter` / `ScrollPresenter` ancestor, which clip without setting `Clip`
- intersect with the `XamlRoot` bounds, covering the "pushed past the app edge" case

`GetBoundingRectangleCore` and `IsOffscreenCore` both read that one helper: offscreen is simply "the clipped rect is empty", which is the same rule the framework uses. Overriding the rect alone was verified to fix the rectangle while still leaving `IsOffscreen = true`, so a rect-only fix looks correct but is not.

## Verification

Same app, same machine, after the change:

| Element | Before | After |
| --- | --- | --- |
| InkCanvas A | `EMPTY/INF`, offscreen=True | `142,369 2400x1800`, offscreen=False |
| InkCanvas B | `EMPTY/INF`, offscreen=True | `1024,295 2400x1800`, offscreen=False |
| InkCanvas (unclipped) | `EMPTY/INF`, offscreen=True | `1775,176 240x66`, offscreen=False |

The values are correct for the display scale: 1600x1200 DIPs x 1.5 = 2400x1800, and 160x44 x 1.5 = 240x66.

Offscreen reporting was checked in both directions. With the canvas positioned outside the window it correctly reports `offscreen=True`; moved inside, it reports `offscreen=False`. During that check an ordinary `Button` pushed off the window edge reported `EMPTY/INF` + `offscreen=True` - the exact signature the visible canvas used to produce.

Six API tests cover the behaviour: reported bounds, offscreen when collapsed or positioned out of the content area, bounds clipped to a `ScrollViewer` viewport (and *not* offscreen while partially visible), offscreen when an ancestor clip excludes the canvas, and offscreen / partially clipped when the clip is set on the canvas itself.

## Also in this PR: InkToolbar localization

`InkToolbar` ships **en-us only**, while every control enrolled with the localization service ships 76-86 locales.

The cause is not the control. Its `.resw` is in the per-control `Strings\en-us` location, `InkToolbar.vcxitems` declares it as a `PRIResource`, and the `SR_` macros match the resource key names - all correct. It was simply never added to the `resourceFilePath` list that the Touchdown task uploads in `build/AzurePipelinesTemplates/WinUI-LocalizeResources-Steps.yml`.

This adds that one line. Only `build/WinUI-Localization.yml` consumes that template, so PR validation is unaffected. The localized `.resw` files are checked in separately once the service returns them.

## Notes

- This is not a regression introduced by the WinUI 3 control. UWP has no `InkCanvasAutomationPeer` at all, and a childless `FrameworkElement` gets no automation peer by default, so the classic control exposes less. This closes an inherited gap.
- A known limitation remains: the core also applies a *layout* clip (`ApplyLayoutClipToBounds`), which the peer does not model. It is not exposed publicly, and approximating it by intersecting every ancestor's bounds would be wrong, because XAML does not clip children to parent bounds by default. The common case is covered by the scroll-viewport intersection. The correct long-term fix is for the element to report layout-sized content bounds in the core, the way `CSwapChainPanel::GenerateContentBounds` already does; that needs a core change and is tracked separately.
- The added API tests could not be run locally: `MUXControlsTestApp` fails to restore `Microsoft.NET.ILLink.Tasks.8.0.31` from the `WinUI-Dependencies` feed with HTTP 401. The behaviour was verified at runtime through a standalone WinUI 3 app instead. Please let CI exercise the tests.

### CODE REVIEW

**Comment by harshit7962** [comment]
> The added API test could not be run locally: MUXControlsTestApp fails to restore Microsoft.NET.ILLink.Tasks.8.0.31 from the WinUI-Dependencies feed with HTTP 401. The behaviour was verified at runtime through a standalone WinUI 3 app instead. Please let CI exercise the test.

@Nitin-100 sir, please let me know if this is needed, I can get it sorted. Could you please let me know what this blocks?

File: controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp [hunk 1]
```diff
@@ -19,3 +20,60 @@ winrt::AutomationControlType InkCanvasAutomationPeer::GetAutomationControlTypeCo
...
+// the layout size instead, matching ColorSpectrum.
+winrt::Rect InkCanvasAutomationPeer::GetBoundingRectangleCore()
+{
+    if (auto const owner = Owner().try_as<winrt::InkCanvas>())
+    {
+        const winrt::Rect localBounds{
+            0.0f,
+            0.0f,
+            static_cast<float>(owner.ActualWidth()),
+            static_cast<float>(owner.ActualHeight()) };
```
Comment (harshit7962) [inline controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp]: Does this account for the fact that the actual width and height can be larger than what is visible on screen? Like if inkcanvas is inside let's say a scroll viewer?

Comment (harshit7962) [reply controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp]: or if lets say the incanvas is beyond the app boundary itself

File: controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp [hunk 2]
```diff
@@ -19,3 +20,60 @@ winrt::AutomationControlType InkCanvasAutomationPeer::GetAutomationControlTypeCo
...
+    }
+
+    // Treat the canvas as offscreen when it is scrolled or positioned outside the content area.
+    const auto bounds = owner.TransformToVisual(nullptr).TransformBounds(localBounds);
+    const auto rootSize = xamlRoot.Size();
+
+    return bounds.X + bounds.Width <= 0.0f
+        || bounds.Y + bounds.Height <= 0.0f
+        || bounds.X >= rootSize.Width
+        || bounds.Y >= rootSize.Height;
```
Comment (harshit7962) [inline controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp]: For my understanding, I hope it covers for the scenario where inkcanvas is partially visible on screen. I am not sure what the value of `IsOffscreenCore` should be, hence double checking the same.

**Comment by Nitin-100** [comment]
> > The added API test could not be run locally: MUXControlsTestApp fails to restore Microsoft.NET.ILLink.Tasks.8.0.31 from the WinUI-Dependencies feed with HTTP 401. The behaviour was verified at runtime through a standalone WinUI 3 app instead. Please let CI exercise the test.
> 
> Nitin Chaudhary (@Nitin-100) sir, please let me know if this is needed, I can get it sorted. Could you please let me know what this blocks?

Thanks Harshit. It doesn't block this PR , PR Build is green and CI will exercise the new test. What it blocked was running it locally before raising, so I verified the behaviour through a standalone WinUI 3 app instead; the before/after numbers in the description are runtime measurements from that.

It would still be worth sorting, since the 401 stops anyone from running MUXControlsTestApp locally, not just this test. The restore fails on Microsoft.NET.ILLink.Tasks.8.0.31 from the WinUI-Dependencies feed with "401 Unauthorized - No local versions of package microsoft.net.illink.tasks; please provide authentication to access versions from upstream that have not yet been saved to your feed." If you can get that package saved to the feed or the auth sorted, that unblocks local ink test runs generally.

File: controls/dev/InkCanvas/APITests/InkCanvasTests.cs [hunk 3]
```diff
@@ -72,6 +73,31 @@ public void InkCanvasInkPresenterAccessorTest()
...
+
+                var peer = FrameworkElementAutomationPeer.CreatePeerForElement(inkCanvas);
+                Verify.IsNotNull(peer, "InkCanvas should create an automation peer.");
+
+                var bounds = peer.GetBoundingRectangle();
+                Log.Comment($"Bounding rectangle: {bounds.X},{bounds.Y} {bounds.Width}x{bounds.Height}");
+
+                Verify.IsGreaterThan(bounds.Width, 0.0, "Bounding rectangle should have a non-zero width.");
+                Verify.IsGreaterThan(bounds.Height, 0.0, "Bounding rectangle should have a non-zero height.");
+                Verify.IsFalse(peer.IsOffscreen(), "A visible InkCanvas should not report itself as offscreen.");
```
Comment (harshit7962) [inline controls/dev/InkCanvas/APITests/InkCanvasTests.cs:98]: I propse we add a tests where this is actually `true`.

File: controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp (hunk 1 above)
Comment (Nitin-100) [reply controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp]: Good catch - it did not, and you are right on both counts.

I checked what the framework itself does. `FrameworkElementAutomationPeer::GetBoundingRectangleCore` and `CFrameworkElementAutomationPeer::IsOffscreenHelper` both go through:

```cpp
GetGlobalBoundsWithOptions(&bounds, false /* ignoreClipping */, ignoreClippingOnScrollContentPresenters, false);
```

so stock peers report **clipped** bounds, and `GetBoundingRectangleCore` returns an empty rect whenever the element is offscreen. My override returned the raw layout rect with no clipping, so an InkCanvas inside a ScrollViewer reported its full content size (I could repro it: a 2400x1800 canvas reported 2400x1800 inside an 838px-tall viewport), and one pushed past the app edge still reported bounds.

Fixed in f6e3593. The bounds are now computed once and intersected with:

- `UIElement.Clip` on any ancestor,
- the viewport of any `ScrollContentPresenter` / `ScrollPresenter` ancestor (these clip without setting `Clip`),
- the `XamlRoot` bounds, which covers the "beyond the app boundary" case.

The layout size is still the starting point, because that part of the original problem is real: InkCanvas renders through a child composition visual and has no XAML children, so `GetGlobalBounds` comes back empty and the default peer reports 0x0.

File: controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp (hunk 2 above)
Comment (Nitin-100) [reply controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp]: It does, and the framework gives us a definition to match rather than guess at.

In `CFrameworkElementAutomationPeer::IsOffscreenHelper`:

```cpp
const bool hasEmptyBounds = (bounds.left == bounds.right && bounds.top == bounds.bottom);
bIsOffscreen = hasEmptyBounds;   // then walks the ancestor chain for IsVisible / IsActive
```

where `bounds` is already clipped. So the rule is: **offscreen means the clipped bounds are empty**. Partially visible therefore means not offscreen, which also matches the UIA docs - `IsOffscreen` is about reachability, not full visibility.

After f6e3593 the peer matches that by construction: both methods now come from one helper that returns the clipped rect, `IsOffscreenCore` is just "is that rect empty", and `GetBoundingRectangleCore` returns an empty rect when it is. So the two cannot disagree.

Worth flagging that this also fixes a case my earlier version got wrong: a canvas scrolled completely out of its ScrollViewer reported `IsOffscreen == false`, because I was only testing against the `XamlRoot` bounds and never the scroll viewport. It now reports `true`.

File: controls/dev/InkCanvas/APITests/InkCanvasTests.cs (hunk 3 above)
Comment (Nitin-100) [reply controls/dev/InkCanvas/APITests/InkCanvasTests.cs:98]: Agreed, added. There are now three tests covering the `true` side:

- `InkCanvasAutomationPeerReportsOffscreen` - a collapsed canvas, and one positioned past the content area, both assert `IsOffscreen() == true`.
- `InkCanvasAutomationPeerReportsOffscreenWhenClippedOut` - canvas is `Visible` and in the tree, but an ancestor clip excludes it; asserts `IsOffscreen() == true` **and** that the rect comes back empty, which is the invariant the framework holds.
- `InkCanvasAutomationPeerClipsBoundsToScrollViewport` - the partially visible case from the other thread: a 400x2000 canvas in a 300x200 ScrollViewer, asserting the reported rect is clipped to the viewport and that it is *not* offscreen.

The viewport test compares the canvas rect against the ScrollViewer's own rect rather than hardcoding pixels, so it does not depend on the display scale the test runs at.

File: controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp [hunk 4]
```diff
@@ -19,3 +20,116 @@ winrt::AutomationControlType InkCanvasAutomationPeer::GetAutomationControlTypeCo
 {
     return winrt::AutomationControlType::Pane;
 }
+
+namespace
```
Comment (harshit7962) [inline controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp]: why are we introducing a `namespace` here?

**Comment by Nitin-100** [comment]
Removed it in b31d447. It was only scoping a file-local `IntersectRects` helper, but that helper should not have existed - the framework already ships `RectHelper::Intersect` (same thing `ScrollView` uses), so the peer now calls that directly and the `namespace` block is gone.

One behavioural note: `RectHelper::Intersect` returns `RectHelper::Empty()` when the rects do not overlap, which is not a zero rect, so the result is normalised to an empty rect before it is returned.

Same commit also fixes the CI break. Inside `namespace Microsoft.UI.Xaml.Tests...` the token `Windows.Foundation.Rect` bound to `Microsoft.Windows.Foundation`, which has no `Rect` - hence `CS0234`. Now imported with a `using Windows.Foundation;` directive like the other API test files and referenced unqualified.

File: controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp (hunk 4 above)
Comment (Nitin-100) [reply controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp]: Removed it in b31d447. It was only scoping a file-local `IntersectRects` helper, but that helper should not have existed - the framework already ships `RectHelper::Intersect` (the same one `ScrollView` uses), so the peer calls that directly now and the `namespace` block is gone.

One behavioural note: `RectHelper::Intersect` returns `RectHelper::Empty()` when the rects do not overlap, which is not a zero rect, so the result is normalised to an empty rect before it is returned.

The same commit also fixes the CI break. Inside `namespace Microsoft.UI.Xaml.Tests...` the token `Windows.Foundation.Rect` bound to `Microsoft.Windows.Foundation`, which has no `Rect` - hence CS0234. It is now imported with a `using Windows.Foundation;` directive like the other API test files and referenced unqualified.

File: controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp [hunk 5]
```diff
@@ -19,3 +20,104 @@ winrt::AutomationControlType InkCanvasAutomationPeer::GetAutomationControlTypeCo
...
+
+        if (ancestor.Visibility() != winrt::Visibility::Visible)
+        {
+            return { 0.0f, 0.0f, 0.0f, 0.0f };
+        }
+
+        if (auto const clip = ancestor.Clip())
+        {
+            bounds = winrt::RectHelper::Intersect(bounds, ancestor.TransformToVisual(nullptr).TransformBounds(clip.Rect()));
+        }
```
Comment (Saharsh979) [inline controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp]: Could we apply the owner’s clip as well? This walk starts at  GetParent(owner) , so  InkCanvas.Clip  is excluded. A canvas with a nonzero layout size and an empty clip still returns its full rectangle and  IsOffscreen == false , even though the clip hides it completely.

The existing negative test puts the clip on a Grid ancestor, so it misses this case. Add equivalent tests with an empty and a partial clip directly on the InkCanvas.

File: controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp [hunk 6]
```diff
@@ -19,3 +20,104 @@ winrt::AutomationControlType InkCanvasAutomationPeer::GetAutomationControlTypeCo
...
+        {
+            continue;
+        }
+
+        if (ancestor.Visibility() != winrt::Visibility::Visible)
+        {
+            return { 0.0f, 0.0f, 0.0f, 0.0f };
+        }
+
+        if (auto const clip = ancestor.Clip())
```
Comment (Saharsh979) [inline controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp]: Should we include the clip geometry’s own transform before converting it to root coordinates? clip.Rect()  is the untransformed rectangle.  ancestor.TransformToVisual(nullptr)  applies the element transform, not clip.Transform .

File: controls/dev/InkCanvas/APITests/InkCanvasTests.cs [hunk 7]
```diff
@@ -72,6 +75,142 @@ public void InkCanvasInkPresenterAccessorTest()
...
+                // Both rectangles come back in the same physical units, so comparing them keeps this
+                // assertion independent of the display scale the test happens to run at.
+                Verify.IsLessThanOrEqual(canvasBounds.Height, scrollBounds.Height + 1.0, "Reported height should be clipped to the scroll viewport.");
+                Verify.IsLessThanOrEqual(canvasBounds.Width, scrollBounds.Width + 1.0, "Reported width should be clipped to the scroll viewport.");
+                Verify.IsGreaterThan(canvasBounds.Height, 0.0, "A partially visible canvas should still report bounds.");
+                Verify.IsFalse(canvasPeer.IsOffscreen(), "A partially visible InkCanvas should not report itself as offscreen.");
+            });
+        }
+
+        // The clip case Harshit asked about: the canvas is inside the tree and visible by its own
```
Comment (Saharsh979) [inline controls/dev/InkCanvas/APITests/InkCanvasTests.cs]: AI SLOP

File: controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp
Comment (Saharsh979) [inline controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp:1]: let's not solve this by substituting the parent’s automation rectangle alone. That cannot account for the child’s own clip or the intermediate clip/transform order.

A stronger approach would be to give the native clipping/transform machinery a layout-sized starting rectangle, instead of its empty render-content bounds. That may require an internal helper/bridge, I am not aware if an existing public API already exposes it, please check. That would help us avoid maintaining a second, incomplete clipping implementation inside InkCanvas.

**Review by Saharsh979 — DISMISSED** [review]

File: controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp (hunk 5 above)
Comment (Nitin-100) [reply controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp]: Fixed in 8871e50. You are right - the walk started at `GetParent(owner)`, so a clip on the canvas itself was never applied and an empty clip still reported the full rectangle with `IsOffscreen == false`.

The canvas and every ancestor now go through one `ApplyElementClip` helper, called for the owner before the walk begins. Added both tests you asked for: `InkCanvasAutomationPeerReportsOffscreenWhenSelfClippedOut` (empty clip directly on the canvas, asserts offscreen and an empty rect) and `InkCanvasAutomationPeerClipsBoundsToOwnClip` (partial clip, asserts the rectangle shrinks but the canvas is not offscreen). The partial test compares against the unclipped rectangle rather than hardcoded pixels so it does not depend on display scale.

File: controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp (hunk 6 above)
Comment (Nitin-100) [reply controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp]: Fixed in 8871e50. The core does honour the geometry transform - `CRectangle::InitializeRectangleClip` applies `pClipGeometry->m_pTransform` to the clip before it is used, with the comment "Don't pass in the transform to ApplyClip - ApplyClip already accounts for the transform on the geometry itself" (dxaml/xcp/core/core/elements/rectangle.cpp). We were not doing that.

`ApplyElementClip` now applies `clip.Transform()` to `clip.Rect()` first, then the element transform to root, matching the core's order.

File: controls/dev/InkCanvas/APITests/InkCanvasTests.cs (hunk 7 above)
Comment (Nitin-100) [reply controls/dev/InkCanvas/APITests/InkCanvasTests.cs]: Fair. Removed in 8871e50 - it described the review thread it came from rather than the test. It now just states the condition under test: "Visible and in the tree, but an ancestor clip excludes it entirely, so it reports offscreen."

I checked the rest of the change for the same thing; that was the only comment written at a reviewer rather than at the reader.

File: controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp
Comment (Nitin-100) [reply controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp:1]: Checked, and the short answer is: no existing public API or bridge exposes this, but your instinct is right and there is an exact precedent in the core.

**The precedent.** `CSwapChainPanel::GenerateContentBounds` does precisely what you describe - it is composition-only content with no XAML children, and it reports the layout size as its content bounds:

```cpp
if (HasContent())
{
    pBounds->left = 0.0f;  pBounds->top = 0.0f;
    pBounds->right  = GetActualWidth();
    pBounds->bottom = GetActualHeight();
}
```

Once that returns a non-empty rectangle, everything downstream - `UIElement.Clip`, layout clips, scroll viewports, ancestor transforms, `XamlRoot` bounds - is handled by the existing machinery, and the peer needs no override at all. That is the right shape for InkCanvas.

**Why it is not reachable from here.** `GenerateContentBounds` is a core C++ virtual on `CUIElement`; the default is `EmptyRectF` (`dxaml/xcp/core/core/elements/uielement.cpp`), reached via `EnsureContentInnerBounds` -> `GetInnerBounds` -> `GetGlobalBoundsOnSelf` -> `GetGlobalBoundsWithOptions`. Only core types override it: `CBorder`, `CPanel` (and only when it has a brush), `CSwapChainPanel`. `InkCanvas` is a MUX control in `Microsoft.UI.Xaml.Controls.dll`, so it cannot.

I also checked the escape hatches you hinted at, and none work:
- `GetGlobalBoundsWithOptions` is not projected to WinRT, and there is no FrameworkUdk-style bridge in `controls/dev` for it.
- `ElementCompositionPreview.SetElementChildVisual` does not contribute to bounds - hand-in visuals are tracked separately (`m_isUsingHandInVisual`) and never reach `GenerateContentBounds`. That is the root cause here.
- `FrameworkElement` has no `Background`, so the `CPanel` brush trick is unavailable without changing the base type.
- `UIElement.Clip` can only restrict bounds, never inflate empty ones.

The only sanctioned extensibility point is `AutomationPeer::GetBoundingRectangle()` bridging to the managed peer, which is what this change uses.

**One correction:** the peer does not substitute the parent''s automation rectangle. It starts from the owner''s own layout size, transforms it to root, then intersects clips. But that does not rescue your broader point - you found two real holes in it (the owner''s own clip, and the geometry transform), which is exactly the failure mode you predicted.

**And a third hole I should flag rather than leave for you to find:** the core also applies a *layout* clip (`ApplyLayoutClipToBounds`, `uielement.cpp`), set by `UpdateLayoutClip` when an element exceeds its layout slot. The peer does not model that. It covers the common case via the `ScrollContentPresenter`/`ScrollPresenter` viewport intersection, but a canvas layout-clipped by a smaller fixed-size parent that sets no `UIElement.Clip` will still be over-reported. I do not think that gap can be closed properly at peer level, which is the argument for doing it your way.

So the options are:
1. Keep the peer-side implementation, with the layout-clip gap documented as a known limitation.
2. Do it like `SwapChainPanel`: have the element report layout-sized content bounds in the core, and delete the peer override entirely.

(2) is the correct fix and would remove the parallel implementation you are objecting to. It needs a core change and a core owner''s sign-off, which is outside what this PR can carry. Happy to take that on as a follow-up - shall I file it and keep this PR to the a11y regression, or would you rather block this until the core change lands?

Comment (Nitin-100) [reply controls/dev/InkCanvas/InkCanvasAutomationPeer.cpp:1]: Following up so this is not left ambiguous. I am resolving this thread and tracking the core change separately, because there is no correct way to close it inside this PR:

- The layout-clip gap cannot be fixed at peer level. `LayoutClipGeometry` is not exposed publicly, and approximating it by intersecting every ancestor''s bounds would be **wrong**, because XAML does not clip children to parent bounds by default (`Canvas`, `Grid` and friends allow overflow). That approximation would under-report, which is the dangerous direction for accessibility.
- Giving the canvas a XAML child to obtain native bounds is also a dead end: content bounds require a real brush (`CPanel::GenerateContentBounds` only produces bounds when `m_pBackground != nullptr`), and a hit-testable transparent fill layered over the canvas would swallow the pointer input inking depends on.

On impact if we ship as-is: this strictly improves the current state rather than trading one bug for another. Today the canvas reports `0x0` and `IsOffscreen == true`, so assistive technology skips it entirely. After this change it reports correctly in every configuration we model; the residual gap only over-reports, and only when an ancestor is fixed-size, smaller than the canvas, sets no `UIElement.Clip`, and is not a scroll presenter. Scroll containers, the common case, are handled. Over-reporting leaves the element reachable with a slightly generous highlight; the current behaviour leaves it invisible.

So the remaining work is the `SwapChainPanel`-style core change, which needs a core owner and its own validation. Reopen or ping me if you would rather hold this PR until that lands - I have no objection to blocking, I just did not want the a11y regression sitting behind a core change by default.

File: build/AzurePipelinesTemplates/WinUI-LocalizeResources-Steps.yml [hunk 8]
```diff
@@ -44,6 +44,7 @@ steps:
         controls\dev\ColorPicker\Strings\en-us\Resources.resw
         controls\dev\CommandBarFlyout\Strings\en-us\Resources.resw
         controls\dev\InfoBar\Strings\en-us\Resources.resw
+        controls\dev\InkToolbar\Strings\en-us\Resources.resw
```
Comment (harshit7962) [inline build/AzurePipelinesTemplates/WinUI-LocalizeResources-Steps.yml:47]: I am not sure if this should be part of it. Since the builds are on for long, I am okay to let this go in.

**Comment by harshit7962** [comment]
image

tests added in this PR are failing the Build. @Nitin-100 sir

File: controls/dev/InkCanvas/InkCanvas.cpp [hunk 9]
```diff
@@ -457,12 +457,21 @@ void InkCanvas::DetachFromVisualLink()
...
-        auto compositor = winrt::CompositionTarget::GetCompositorForCurrentThread();
-        // GetForSystemEngine takes any composition object (IInspectable); pass the compositor
-        // directly rather than allocating a throwaway visual just to probe the engine.
-        // CompositionEngine lives in the Microsoft.UI.Composition namespace (it was promoted out of
-        // the Experimental namespace in the InteractiveExperiences transport), so reference it there.
-        return winrt::Microsoft::UI::Composition::CompositionEngine::GetForSystemEngine(compositor) != nullptr;
+        // CompositionEngine is not activatable on every OS build; there GetForSystemEngine throws
+        // CLASS_E_CLASSNOTAVAILABLE, and the lifted path still renders ink.
+        try
+        {
```
Comment (harshit7962) [inline controls/dev/InkCanvas/InkCanvas.cpp:463]: is it that the control will not be available for all the OSes?

Comment (Nitin-100) [reply controls/dev/InkCanvas/InkCanvas.cpp:463]: No

Comment (harshit7962) [reply controls/dev/InkCanvas/InkCanvas.cpp:463]: sorry, is it No for the question or no for the fact that it will not be available across all Oses?

**Review by Saharsh979 — APPROVED** [review]

File: controls/dev/InkCanvas/InkCanvas.cpp (hunk 9 above)
Comment (Nitin-100) [reply controls/dev/InkCanvas/InkCanvas.cpp:463]: It is not available for all OS versions.

**Review by harshit7962 — APPROVED** [review]
